cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocClient.java

```diff
@@ -122,6 +130,41 @@ public static DataprocClient fromConf(DataprocConf conf) throws IOException, Gen
     String subnet = conf.getSubnet();
     Network networkInfo = getNetworkInfo(projectId, network, compute);
 
+    PeeringState state = peeringState(systemNetwork, systemProjectId, networkInfo, compute);
+
+    if (privateInstance) {
+      // private instance must have a peering established
+      verifyPeering(conf, network, systemNetwork, projectId, systemProjectId, state);
+      if (conf.isPreferExternalIP()) {
+        // When prefer external IP is set to true it means only Dataproc external ip can be used to for communication
+        // the instance being private instance is incapable of using external ip for communication
+        throw new IllegalArgumentException("The instance is incapable of using external ip for communication " +
```

Comment:
Why this has to be an exception? Shouldn't we just ignore the prefer external IP config if it is private instance?

Reply:
It's ok to log a warning, but seems unnecessary to be an exception.

Reply:
IIUC from last conversations with Albert preferExternal is defined as if set to true externalIP must be used for communication and if it false then provisioner can pick external/internal according to setup. 

https://github.com/cdapio/cdap/pull/11457#discussion_r301690124

Reply:
the default is false, which means this happens if the user explicitly sets it to true. In this scenario, I think it makes more sense to error rather than ignore a setting they did.

Ideally the setting doesn't even show up in the UI if the provisioner is configured to require private clusters. We discussed in person how we might do this, but I believe it was decided to implement that in a later change.

This should be triggered when the profile is being created, not at runtime.

Reply:
Hm actually there is an issue checking this at create time because macro evaluation of the service account key won't be performed at that time... ignore the part about checking this at create time for now.

Reply:
Done.